### PR TITLE
Placeholder fields and Osquery UI updates

### PR DESCRIPTION
**Pull request by nastasha-solomon**
Fixes https://github.com/elastic/security-docs/issues/2981.

Previews:
- [Use placeholder fields in Osquery queries](https://security-docs_3045.docs-preview.app.elstc.co/guide/en/security/master/osquery-placeholder-fields.html) - New topic
- [Add Osquery Response Actions to rules](https://security-docs_3045.docs-preview.app.elstc.co/guide/en/security/master/osquery-response-action.html#add-osquery-response-action) - Added note under step 2. 
- [Add live queries to an investigation guide](https://security-docs_3045.docs-preview.app.elstc.co/guide/en/security/master/invest-guide-run-osquery.html#add-live-queries-ig) - Added same note under step 3. 
- [Run Osquery from alerts](https://security-docs_3045.docs-preview.app.elstc.co/guide/en/security/master/alerts-run-osquery.html) - Added same note under step 4.

### CODE REVIEW

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[[osquery-placeholder-fields]]
+== Use placeholder fields in Osquery queries 
+
+Placeholder fields dynamically inject alert and event data into queries that you run. To create a placeholder field, you must wrap the field in double curly brackets (`{{}}`); you can use any field in an event or alert document.
```
Comment (nastasha-solomon) [inline docs/osquery/osquery-placeholder-fields.asciidoc:4]:
```suggestion
Placeholder fields dynamically inject alert and event data into Osquery queries at runtime. Placeholder fields use mustache syntax and must be wrapped in double curly brackets (`{{}}`); you can turn any field in an event or alert document into a placeholder field. 
```

**Comment by szwarckonrad** [comment]
CC @tomsonpl @patrykkopycinski

File: docs/osquery/osquery-index.asciidoc [hunk 2]
```diff
@@ -6,4 +6,6 @@ include::invest-guide-run-osquery.asciidoc[][leveloffset=+1]
 
 include::alerts-run-osquery.asciidoc[][leveloffset=+1]
 
-include::view-osquery-results.asciidoc[][leveloffset=+1]
\ No newline at end of file
+include::view-osquery-results.asciidoc[][leveloffset=+1]
```
Comment (szwarckonrad) [inline docs/osquery/osquery-index.asciidoc:9]: Comment test.

**Review by tomsonpl — APPROVED** [review]
LGTM 👍

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+* <<osquery-response-action,Osquery Response Action>>
++
+NOTE: Placeholder fields within an Osquery Response Action query will only work if the rule has generated alerts.
+* <<invest-guide-run-osquery,Investigation guides using Osquery queries>> 
+
+[float]
+[[placeholder-field-syntax]]
+=== Placeholder field syntax and requirements
+
+Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{}}`). You can use any field within an event or alert document as a placeholder field.
```
Comment (joepeeples) [inline docs/osquery/osquery-placeholder-fields.asciidoc:18]: All those brackets are a little hard to parse visually. How about something like:

```suggestion
Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{example.field}}`). You can use any field within an event or alert document as a placeholder field. 
```

**Review by joepeeples — COMMENTED** [review]
Good stuff! Just added a couple small suggestions, thanks!

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 4]
```diff
@@ -0,0 +1,30 @@
+[[osquery-placeholder-fields]]
+== Use placeholder fields in Osquery queries 
+
+Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like replaceable parameters. You can use placeholder fields to build flexible and re-usable queries.
```
[the remark this answers is not in the document]
Comment (benironside) [reply docs/osquery/osquery-placeholder-fields.asciidoc:4]: I think maybe it would make more sense without the word "replaceable". I think it's redundant because a parameter is a variable that helps define some kind of function (in this case a query), so the name of the parameter in the query definition is meant to be replaced by some value, which I think is what this feature does. (I.e. you follow the syntax explained below to define how the parameter gets its value, and then the query runs with that value. )

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
+[[osquery-placeholder-fields]]
+== Use placeholder fields in Osquery queries 
+
+Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like replaceable parameters. You can use placeholder fields to build flexible and re-usable queries. 
+
+Placeholder fields work in single queries or query packs. They are also supported in the following features:
+
+* <<alerts-run-osquery,Live queries>>
+* <<osquery-response-action,Osquery Response Action>>
```
Comment (benironside) [inline docs/osquery/osquery-placeholder-fields.asciidoc:9]:
```suggestion
* <<osquery-response-action,Osquery Response Actions>>
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
++
+NOTE: Placeholder fields within an Osquery Response Action query will only work if the rule has generated alerts.
```
Comment (benironside) [inline docs/osquery/osquery-placeholder-fields.asciidoc:11]: Not sure I understand the purpose of this note. I could be missing something, but why would someone expect a response action to run if it's parent rule hadn't generated an alert?

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+* <<invest-guide-run-osquery,Investigation guides using Osquery queries>> 
+
+[float]
+[[placeholder-field-syntax]]
+=== Placeholder field syntax and requirements
+
+Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{}}`). You can use any field within an event or alert document as a placeholder field. 
+
+Queries with placeholder fields _must_ be run against an alert or event. If a placeholder field can't find an alert or event value, the query status will be `Skipped`.
```
Comment (benironside) [inline docs/osquery/osquery-placeholder-fields.asciidoc:20]:
```suggestion
Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and be `Skipped`. 
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+[float]
+[[placeholder-field-example]]
+==== Example query with a placeholder field
+
+In the following query, the `{{host.name}}` placeholder field is used: 
+
+`SELECT * FROM os_version WHERE name = `{{host.name}}``
+
+When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.name}}` placeholder field.
```
Comment (benironside) [inline docs/osquery/osquery-placeholder-fields.asciidoc:30]: very clear example, great illustration

File: docs/osquery/osquery-response-action.asciidoc [hunk 9]
```diff
@@ -27,6 +27,9 @@ You can add Osquery Response Actions to new or existing custom query rules. Quer
 ** *Existing rule*: Edit the rule's settings, then go to the *Actions* tab. In the tab, click the *osquery* icon under the Response Actions section.
 . Specify whether you want to set up a single live query or a pack:
 ** *Query*: Select a saved query or enter a new one. After you enter the query, you can expand the **Advanced** section to view or set {kibana-ref}/osquery.html#osquery-map-fields[mapped ECS fields] included in the results from the live query. Mapping ECS fields is optional.
++
+TIP: Use <<osquery-placeholder-fields,placeholder fields>> to dynamically add existing alert data to your query.
```
Comment (benironside) [inline docs/osquery/osquery-response-action.asciidoc:31]:
```suggestion
TIP: You can use <<osquery-placeholder-fields,placeholder fields>> to dynamically add alert data to your query. 
```

**Review by benironside — COMMENTED** [review]
Looks excellent overall, good examples and I think I could use this feature no problem after reading this — I left a few small suggestions for you to consider

File: docs/osquery/osquery-placeholder-fields.asciidoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (nastasha-solomon) [reply docs/osquery/osquery-placeholder-fields.asciidoc:4]:
```suggestion
Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and re-usable queries. 
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 10]
```diff
@@ -0,0 +1,29 @@
+[[osquery-placeholder-fields]]
+== Use placeholder fields in Osquery queries 
+
+Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and re-usable queries.
```
Comment (jmikell821) [inline docs/osquery/osquery-placeholder-fields.asciidoc:4]:
```suggestion
Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and reusable queries. 
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+[float]
+[[placeholder-field-example]]
+==== Example query with a placeholder field
+
+The following query uses the `{{host.name}}` placeholder field: 
+
+`SELECT * FROM os_version WHERE name = `{{host.name}}``
+
+When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.name}}` placeholder field.
```
Comment (jmikell821) [inline docs/osquery/osquery-placeholder-fields.asciidoc:29]: `passed into` sounds a little strange. Would `transferred` work here? 
```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be transferred to the `{{host.name}}` placeholder field. 
```

**Review by jmikell821 — COMMENTED** [review]
Two small comments, otherwise, looks good!

File: docs/osquery/osquery-placeholder-fields.asciidoc (hunk 7 above)
Comment (nastasha-solomon) [reply docs/osquery/osquery-placeholder-fields.asciidoc:19]: @benironside I'm fairly certain that the second sentence needs to be kept as is. If an Osquery query is using a placeholder field to retrieve a field that doesn't exist on the alert or event, the query will fail regardless of where it was ran against. @szwarckonrad @tomsonpl can you verify what I said is correct? 

@szwarckonrad @tomsonpl - On a related note, during my testing, I've seen the `error` status returned for failed queries (i.e. queries using placeholder fields to pass in values for fields that don't exist on an alert or event). Is this correct? Should I replace `Skipped` with `error`?

File: docs/osquery/osquery-placeholder-fields.asciidoc (hunk 8 above)
Comment (nastasha-solomon) [reply docs/osquery/osquery-placeholder-fields.asciidoc:29]:
```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.os.name}}` placeholder field. 
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+[float]
+[[placeholder-field-example]]
+==== Example query with a placeholder field
+
+The following query uses the `{{host.name}}` placeholder field: 
+
+`SELECT * FROM os_version WHERE name = `{{host.name}}``
+
+When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.os.name}}` placeholder field.
```
Comment (nastasha-solomon) [inline docs/osquery/osquery-placeholder-fields.asciidoc:29]:
```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be transferred to the `{{host.os.name}}` placeholder field. 
```

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+Queries with placeholder fields _must_ be run against an alert or event. If a placeholder field can't find an alert or event value, the query status will be `Skipped`. 
+
+[float]
+[[placeholder-field-example]]
+==== Example query with a placeholder field
+
+The following query uses the `{{host.name}}` placeholder field: 
+
+`SELECT * FROM os_version WHERE name = `{{host.name}}``
```
Comment (nastasha-solomon) [inline docs/osquery/osquery-placeholder-fields.asciidoc:27]:
```suggestion
`SELECT * FROM os_version WHERE name = `{{host.os.name}}``
```

File: docs/osquery/osquery-placeholder-fields.asciidoc (hunk 7 above)
Comment (tomsonpl) [reply docs/osquery/osquery-placeholder-fields.asciidoc:19]: Hey @nastasha-solomon , @benironside - to me both of your version seem the same (correct) :P 

Whenever we use the {{ data }} dynamic params syntax within context of Alert or Event (we need its data) - we try to replace the parameter with the data, and if we fail (eg. there is no data held under the key we provided) - we return an `error` saying "This query hasn't been called due to parameter used and its value not found in the alert." which means that in the end we didn't run this osquery on the host (not sending action to fleet). To me this can safely be referred to as something like "The query was skipped" 

I am sorry as this might sound confusing, but I'll be happy to discuss this further if that would help 👍

Comment (nastasha-solomon) [reply docs/osquery/osquery-placeholder-fields.asciidoc:19]: Ah, ok thanks for the additional information about the error message and how Fleet perceives (or doesn't perceive?) the "failed" query. With this in mind, I agree that @benironside's edit makes sense to use. Thanks, both!

File: docs/osquery/osquery-placeholder-fields.asciidoc [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
++
+* <<invest-guide-run-osquery,Investigation guides using Osquery queries>> 
+
+[float]
+[[placeholder-field-syntax]]
+=== Placeholder field syntax and requirements
+
+Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{example.field}}`). You can use any field within an event or alert document as a placeholder field. 
+
+Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and be `Skipped`.
```
Comment (nastasha-solomon) [inline docs/osquery/osquery-placeholder-fields.asciidoc:19]:
```suggestion
Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and the query status will be `error`. 
```
